File: Provenance/Controller/PVNESControllerViewController.m
//
//  PVNESControllerViewController.m
//  Provenance
//
//  Created by James Addyman on 22/03/2015.
//  Copyright (c) 2015 James Addyman. All rights reserved.
//

#import "PVNESControllerViewController.h"
#import <PVFCEU/PVFCEUEmulatorCore.h>

@interface PVNESControllerViewController ()

@end

@implementation PVNESControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue;
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"A"])
        {
            [button setTag:PVNESButtonA];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"B"])
        {
            [button setTag:PVNESButtonB];
        }
    }
    
    [self.startButton setTag:PVNESButtonStart];
    [self.selectButton setTag:PVNESButtonSelect];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    
    [nesCore releaseNESButton:PVNESButtonUp forPlayer:0];
    [nesCore releaseNESButton:PVNESButtonDown forPlayer:0];
    [nesCore releaseNESButton:PVNESButtonLeft forPlayer:0];
    [nesCore releaseNESButton:PVNESButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [nesCore pushNESButton:PVNESButtonUp forPlayer:0];
            [nesCore pushNESButton:PVNESButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [nesCore pushNESButton:PVNESButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [nesCore pushNESButton:PVNESButtonUp forPlayer:0];
            [nesCore pushNESButton:PVNESButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [nesCore pushNESButton:PVNESButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [nesCore pushNESButton:PVNESButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [nesCore pushNESButton:PVNESButtonDown forPlayer:0];
            [nesCore pushNESButton:PVNESButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [nesCore pushNESButton:PVNESButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [nesCore pushNESButton:PVNESButtonDown forPlayer:0];
            [nesCore pushNESButton:PVNESButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
	[super dPad:dPad didPressDirection:direction];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    
    [nesCore releaseNESButton:PVNESButtonUp forPlayer:0];
    [nesCore releaseNESButton:PVNESButtonDown forPlayer:0];
    [nesCore releaseNESButton:PVNESButtonLeft forPlayer:0];
    [nesCore releaseNESButton:PVNESButtonRight forPlayer:0];
	[super dPadDidReleaseDirection:dPad];
}

- (void)buttonPressed:(JSButton *)button
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    [nesCore pushNESButton:[button tag] forPlayer:0];
    
	[super buttonPressed:button];
}

- (void)buttonReleased:(JSButton *)button
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    [nesCore releaseNESButton:[button tag] forPlayer:0];
	[super buttonReleased:button];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    [nesCore pushNESButton:PVNESButtonStart forPlayer:player];
	[super pressStartForPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    [nesCore releaseNESButton:PVNESButtonStart forPlayer:player];
	[super releaseStartForPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    [nesCore pushNESButton:PVNESButtonSelect forPlayer:player];
	[super pressSelectForPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    PVFCEUEmulatorCore *nesCore = (PVFCEUEmulatorCore *)self.emulatorCore;
    [nesCore releaseNESButton:PVNESButtonSelect forPlayer:player];
	[super releaseSelectForPlayer:player];
}

@end

File: Provenance/Controller/PVNESControllerViewController.swift
//
//  PVNESControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright © 2018 Joe Mattiello. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVNESButton {
        get {
            return PVNESButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVNESControllerViewController: PVControllerViewController<PVNESSystemResponderClient> {
    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let title = button.titleLabel?.text else {
                return
            }
            if title == "A" {
                button.buttonTag = .a
            } else if title == "B" {
                button.buttonTag = .b
            }
        }

        startButton?.buttonTag = .start
        selectButton?.buttonTag = .select
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.up, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.left, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.right, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.down, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            default:
                break
        }
        super.dPad(dPad, didPress: direction)
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        super.dPadDidReleaseDirection(dPad)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        super.buttonPressed(button)
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
        super.buttonReleased(button)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.start, forPlayer: player)
        super.pressStart(forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.start, forPlayer: player)
        super.releaseStart(forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.select, forPlayer: player)
        super.pressSelect(forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.select, forPlayer: player)
        super.releaseSelect(forPlayer: player)
    }
}
